### PR TITLE
HoS gun beacon: make your choice from the X-01, X-02 and X-03 guns

### PR DESCRIPTION
**Pull request by VictorVow**
## About The Pull Request

Revived Cae's original [X-02 PR](https://github.com/BeeStation/BeeStation-Hornet/pull/2216) with some tweaks. 

Swaps the HoS gun for a beacon which allows you to choose your gun, includes the original X-01 as well as the new X-02 and X-03 experimental models. Allows you to choose a gun based on your needs, whether it's for high utility in a small amount of space, high impact ballistics or the versatility of the existing X-01.

### X-02 MultiPhase Energy Gun
Energy-based gun focused around utility ammo. Lacks lethal ammo. 
- Ion: Exactly the same as the ion rifle rounds. 12 shots, 3 shots/second.
- Freeze: Same as the temperature gun's only useful setting, but with a lower rate of fire. 20 shots, 3 shots/second.
- Netting: Same as DRAGnet netting i.e. the spreading shots. 12 shots, 2 shots/second. 

### X-03 MultiPhase Hybrid Gun
Hybrid ballistics created from an internal 3D printer. Lacks non-lethals and most utility ammo. 
- .454 High Velocity: Does 25 brute damage per shot with low AP and moves significantly faster than regular projectiles. 10 shots, 2 shots/second.
- .454 Anti Riot/Plastic: Semi-lethal, does 5 brute and 15 stamina damage per pellet with 3 pellets per shot. 12 shots, 4 shots/second.
- .454 Breaching: Renamed 12g breaching slugs. Good for busting down doors and windows, terrible at everything else. 12 shots, 3 shots/second.

### Why It's Good For The Game

More interesting variety by being able to switch up your gun from round-to-round. Encourages using a diversity of ammo you might otherwise ignore. Helps you to regret which type you chose every time you enter deadchat. 

## Changelog
:cl:
add: Adds the HoS gun beacon. Make your choice from the X-01, X-02 and X-03 guns.
balance: HoS gun now uses a mindshield firing pin. 
/:cl:

### CODE REVIEW

**Comment by TeomanTheGreat** [comment]
This doesn't mess with the tot objective right?

**Comment by VictorVow** [comment]
Yeah that's fine.

**Comment by PowerfulBacon** [comment]
Why does an energy gun shoot shotgun pellets

**Comment by ItsTisme** [comment]
> Why does an energy gun shoot shotgun pellets

bluespace

**Comment by AnjelicaChaykovsky** [comment]
@Victor239  what about making different sprites for x03 and x02

ballistic model could be for example red and cool 

i mean it make sense that it would look different since laser and ballistic need different parts 

i  could finish this sprite and make x02 too

**Comment by Cenrus** [comment]
please get rid of the tazer mode.

**Comment by Moccha-Bee** [comment]
> please get rid of the tazer mode.

why?

**Comment by AnjelicaChaykovsky** [comment]
> > please get rid of the tazer mode.
> 
> why?

its pretty shit and   each shot is 33% ammo

**Comment by Moccha-Bee** [comment]
> > > please get rid of the tazer mode.
> > 
> > 
> > why?
> 
> its pretty shit and each shot is 33% ammo

So what, i like using the taser mode to insta stun the criminal scum 😎

**Comment by AnjelicaChaykovsky** [comment]
> So what, i like using the taser mode to insta stun the criminal scum 😎
 then they get up from that knockdown half a second later

**Comment by Moccha-Bee** [comment]
> > So what, i like using the taser mode to insta stun the criminal scum 😎
> > then they get up from that knockdown half a second later

the taser stun lasts longer, and i just disabler/stunbaton them afterwards. Its great way to non-lethal way to stopp people who are fleeing

**Comment by Cenrus** [comment]
25 damage from the HoS gun is far too much considering Stetchkins do 20 and need to be reloaded with real ammo that can't be recharged.

**Comment by AnjelicaChaykovsky** [comment]
> 25 damage from the HoS gun is far too much considering Stetchkins do 20 and need to be reloaded with real ammo that can't be recharged.

(looks at roundstart riot shotgun that can hold shotgun slugs and crit people in two shots with great accuracy)

**Comment by VictorVow** [comment]
@PowerfulBacon energy to matter conversion is prevalent on station, from chem dispensers to bluespace miners. A 3D printer for ammo is no different. 

@AnjelicaChaykovsky different sprites sounds good, I'll probably bug Azlanon for that though. 

Taser mode - it's short-range and high energy cost make it situational, even then shotguns are comparably good and better if examining the warden's combat shotgun even. I might reduce the fire rate for that mode however. 

@Cenrus when choosing the ammo for the X-03 the warden's shotgun was my main point of comparison, and in this their shotgun is still stronger however it lacks the versatility of the X-03.

**Comment by MrHorizons** [comment]
I'd say you should rename the X-03 to the X-03 MultiPhase Ballistic Gun, since it's only, well, ballistics.

**Comment by WhyIsCaeciliusTaken** [comment]
@cenrus the stechkin does 30. The detective's revolver does 25 and also has the same rate of fire as this new gun. Only difference is the faster travelling bullets.

**Comment by AnjelicaChaykovsky** [comment]
> @AnjelicaChaykovsky different sprites sounds good, I'll probably bug Azlanon for that though.

ok rest in peace my effort

**Comment by utilbird** [comment]
We should be removing non-antag ballistics, not adding more 😔

**Comment by Blueturbo47** [comment]
Don't you need to change the trator objective too? If a tot has to steal the gun from the hos and it's already changed they won't be able to greentext.

**Comment by utilbird** [comment]
@Blueturbo47 it won’t

**Comment by ZombiesVsPlants** [comment]
>round start ion gun with alternate settings
>shoots 3 bolts per second
>fits in backpacks

Rip malf/nukie borgs

**Comment by eeSPee** [comment]
You could have the ballistic one recharge with Iron, so it can't print ammo out of raw energy.

**Comment by utilbird** [comment]
@AnjelicaChaykovsky the ballistic gun looks like something from evangallion

**Comment by BrotherHangyul** [comment]
Me likey. Maybe make the X-03 sprite a bit more revolver/big-irony.

**Comment by ToasterBan** [comment]
i hate sec and i still love this

**Review by BrotherHangyul — COMMENTED** [review]
Seems like it would work fine and add new content to the game.

File: code/game/objects/items/miscellaneous.dm [hunk 1]
```diff
@@ -63,6 +63,22 @@
...
+
+/obj/item/choice_beacon/hosgun/generate_display_names()
+	var/static/list/hosgun_item_list
+	if(!hosgun_item_list)
+		hosgun_item_list = list()
+		var/list/templist = typesof(/obj/item/gun/energy/e_gun/hos)
+		for(var/V in templist)
+			var/atom/A = V
+			hosgun_item_list[initial(A.name)] = A
+	return hosgun_item_list
```
Comment (ike709) [inline code/game/objects/items/miscellaneous.dm]: Bad code is bad.

**Comment by ike709** [comment]
I'll only consider this PR if the guns are DNA-locked to the HoS so a tider can't robust them and steal it.

**Comment by VictorVow** [comment]
I'd hesitate to add a mindshield firing pin let alone a DNA locked firing pin for the HOS gun, and I say this as a security main. 

Killing the HOS or robbing their locker is a time honoured tradition, as well as an antag objective. Firing pins are quite hard to get a hold of in addition to obtaining their gun. For 95% of antags who obtain the gun it'll be unusable for them because they lack the pin. 

Even the warden's combat shotgun and captain's antique laser gun use standard firing pins so I don't see the reasoning for such an antag nerf.

**Comment by BrotherHangyul** [comment]
Maybe make the gun emaggable?

**Comment by TeomanTheGreat** [comment]
> Maybe make the gun emaggable?

Why would it need to be emaggable?

**Comment by BrotherHangyul** [comment]
merge

File: code/game/objects/items/miscellaneous.dm (hunk 1 above)
Comment (VictorVow) [reply code/game/objects/items/miscellaneous.dm]: It's a copy and paste of the other choice beacons

**Comment by VictorVow** [comment]
@park66665 review instead of marking stale

**Comment by park66665** [comment]
Did you convince ike?

**Comment by VictorVow** [comment]
@ike709 see my reply to DNA lock feedback

File: code/game/objects/items/miscellaneous.dm (hunk 1 above)
Comment (ike709) [reply code/game/objects/items/miscellaneous.dm]: Doesn't make it less bad.

**Comment by ike709** [comment]
I'll compromise: Mindshield firing pin.

**Comment by Cenrus** [comment]
Why should a traitor not be allowed to use the gun they put in so much effort to get their hands on? Or at least, not without a mindshield which is arguably even harder to obtain.

File: code/modules/projectiles/ammunition/energy/special.dm [hunk 2]
```diff
@@ -65,3 +65,43 @@
...
+/obj/item/ammo_casing/energy/hos/hv
+	fire_sound = 'sound/weapons/rifleshot.ogg'
+	e_cost = 120
+	select_name = ".454HV"
+	projectile_type = /obj/item/projectile/bullet/hos/hv
+
+/obj/item/ammo_casing/energy/hos/trac
+	e_cost = 400
+	select_name = ".454TRAC"
+	projectile_type = /obj/item/projectile/bullet/c38/trac/hos
```
Comment (park66665) [inline code/modules/projectiles/ammunition/energy/special.dm]: Currently this thing is unused. Do you have any plan to use it?

File: code/modules/projectiles/projectile/bullets/special.dm [hunk 3]
```diff
@@ -25,3 +25,25 @@
...
+	name = "3D printed .454 round"
+	damage = 25
+
+/obj/item/projectile/bullet/hos/hv
+	name = "3D printed .454 HV round"
+	speed = 0.2
+	armour_penetration = 10
+
+/obj/item/projectile/bullet/c38/trac/hos
+	name = "3D printed .454 TRAC"
```
Comment (park66665) [inline code/modules/projectiles/projectile/bullets/special.dm]: ditto

File: code/modules/projectiles/ammunition/energy/special.dm (hunk 2 above)
Comment (VictorVow) [reply code/modules/projectiles/ammunition/energy/special.dm]: Can't recall if it was from the original PR or a half-finished idea. Removing it for now anyway.

**Comment by VictorVow** [comment]
Added the requested mindshield firing pin

File: code/game/objects/items/miscellaneous.dm [hunk 4]
```diff
@@ -63,6 +63,22 @@
 
 	new /obj/effect/DPtarget(get_turf(src), pod)
 
+//HoS gun beacon
+/obj/item/choice_beacon/hosgun
+	name = "Head of Security gun beacon"
+	desc = "Choose one of several experimental guns created by leading scientists at NT to put it through proper field testing."
+
+/obj/item/choice_beacon/hosgun/generate_display_names()
+	var/static/list/hosgun_item_list
```
Comment (ike709) [inline code/game/objects/items/miscellaneous.dm]: I don't see why this needs to be static. There should only ever be one HoS gun beacon so you can just store the list as a regular list in the type definiton.

File: code/game/objects/items/miscellaneous.dm (hunk 1 above)
Comment (ike709) [reply code/game/objects/items/miscellaneous.dm]: I'm willing to put this PR to a vote but it won't be merged until after you've cleaned up this proc.

File: code/game/objects/items/miscellaneous.dm (hunk 4 above)
Comment (VictorVow) [reply code/game/objects/items/miscellaneous.dm]: Fixed

**Comment by utilbird** [comment]
Key turned, overall maintainer verdict voted in favour this PR being closed. Feel free to re-open after getting approval from a maintainer, preferably after some balance changes are made

**Comment by VictorVow** [comment]
What balance changes? I already made all the ones Ike requested.

**Comment by ike709** [comment]
🔑

**Pull request closed without merging**